clients/adapter-amqp/src/main/java/org/eclipse/hono/adapter/client/amqp/RequestResponseClient.java

```diff
@@ -0,0 +1,647 @@
+     *
+     * @param subject The subject system property of the message.
+     * @param address The address of the message, put in the <em>to</em> property.
+     * @param appProperties The map containing arbitrary application properties.
+     *                      Maybe null if no application properties are needed.
+     * @return The message constructed from the provided parameters.
+     * @throws NullPointerException if the subject is {@code null}.
+     * @throws IllegalArgumentException if the application properties contain values of types that are not
+     *                                  {@linkplain AbstractHonoClient#setApplicationProperties(Message, Map)
+     *                                  supported by AMQP 1.0}.
+     */
+    private Message createMessage(
+            final String subject,
+            final String address,
+            final Map<String, Object> appProperties) {
+
+        Objects.requireNonNull(subject);
+        final Message msg = ProtonHelper.message();
+        AbstractHonoClient.setApplicationProperties(msg, appProperties);
+        msg.setAddress(address);
+        msg.setReplyTo(replyToAddress);
+        msg.setMessageId(createMessageId());
+        msg.setSubject(subject);
+        return msg;
+    }
+
+    /**
+     * Creates a request message for a payload and headers and sends it to the peer.
+     * <p>
+     * This method uses the sender link's target address as the value for the message's
+     * <em>to</em> property.
+     * <p>
+     * This method first checks if the sender has any credit left. If not, a failed future is returned immediately.
+     * Otherwise, the request message is sent and a timer is started which fails the returned future,
+     * if no response is received within <em>requestTimeoutMillis</em> milliseconds.
+     * <p>
+     * In case of an error the {@code Tags.HTTP_STATUS} tag of the span is set accordingly.
+     * However, the span is never finished by this method.
+     *
+     * @param action The operation that the request is supposed to trigger/invoke.
+     * @param properties The headers to include in the request message as AMQP application properties.
+     * @param payload The payload to include in the request message as an AMQP Value section.
+     * @param contentType The content type of the payload.
+     * @param responseMapper A function mapping a raw AMQP message to the response type.
+     * @param currentSpan The <em>Opentracing</em> span used to trace the request execution.
+     * @return A future indicating the outcome of the operation.
+     *         The future will be failed with a {@link ServerErrorException} if the request cannot be sent to
+     *         the remote service, e.g. because there is no connection to the service or there are no credits
+     *         available for sending the request or the request timed out.
+     * @throws NullPointerException if any of action, result handler or currentSpan is {@code null}.
```

Comment:
No `result handler` param here. It looks like the `responseMapper` param mustn't be null instead here.